# Recipe Recommendation System 
***THE BOMA POT***

### **Author:** ***DS Martin Waweru***

## Table Of Contents

- Business Understanding
- Data Understanding
- Data Preparation
- Modelling 
- Evaluation 
  

## Business Understanding

### Business Description 
The Boma Pot is a web-based culinary platform dedicated to sharing authentic African recipes with the world. Our mission is to enhance the cooking experience of home chefs by providing them with a diverse collection of recipes rooted in African culinary traditions, while also incorporating global influences. Whether you're an experienced cook or just starting, Tasty Dishes offers a wide variety of recipes that empower users to create delicious, flavorful meals from the comfort of their homes.


## Business Goal 
### Objective
The main objective of this project is to develop an item-based recipe recommendation system that suggests recipes to users based on the ingredients they have available. By analyzing the ingredients present in various recipes, the system aims to provide relevant and appealing recommendations that encourage users to explore and cook diverse dishes rooted in African culinary traditions, while also incorporating global flavors.

### Scope

1. Ingredient-Based Recommendations: Develop an algorithm that analyzes user-provided ingredients to recommend recipes based on ingredient similarity, leveraging a diverse dataset that includes recipe_Title, Ingredients, and Instructions for authentic African and global dishes.

2. User-Friendly Interface: Design an intuitive web interface that enables users to input their available ingredients and view tailored recipe recommendations, along with detailed cooking instructions and a feedback mechanism to enhance recommendation accuracy.

## Data Understanding

### Data Source:

We are working with two datasets. The first is from the Food and Agriculture Organization of the United Nations, in collaboration with Kenya's Ministry of Agriculture and Livestock Development. Originally in PDF format, the necessary information was extracted and converted into a CSV file. The second dataset comes from Kaggle, titled the Food Ingredients and Recipe Dataset with Image Name Mapping, which provides global recipe data.

### Access data
click this link below:

1. [Ministry of Agriculture](https://kilimo.go.ke/wp-content/uploads/2021/01/KENYA-FOOD-RECIPES-2018.pdf)
2. [Food Ingredients and Recipe Dataset](https://www.kaggle.com/datasets/pes12017000148/food-ingredients-and-recipe-dataset-with-images)

In [8]:
# Necessary Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import string
import re
import nltk
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


## Data Preparation

### Data Frame One
The dataset contains a CSV file and a zipped folder, consisting of 13,501 rows  and 5 columns respectively as follows:
1. Title: Represents the Title of the Food Dish.
2. Ingredients: Contains the ingredients as they were scraped from the website.
3. Instructions: Has the recipe instructions to be followed to recreate the dish.
4. Image_Name: Has the name of the image as stored in the Food Images zipped folder.
5. Cleaned_Ingredients: Contains the ingredients after being processed and cleaned.

In [9]:
# Load the dataframe

df = pd.read_csv("Data\Food Ingredients and Recipe Dataset with Image Name Mapping.csv", index_col=0)


# Display the first columns
display(df.head(10))

#show the shape
print(df.shape)

Title  \
0  Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                    Crispy Salt and Pepper Potatoes   
2                        Thanksgiving Mac and Cheese   
3                 Italian Sausage and Bread Stuffing   
4                                       Newton's Law   
5                                       Warm Comfort   
6                                 Apples and Oranges   
7                                 Turmeric Hot Toddy   
8                            Instant Pot Lamb Haleem   
9     Spiced Lentil and Caramelized Onion Baked Eggs   

                                         Ingredients  \
0  ['1 (3½–4-lb.) whole chicken', '2¾ tsp. kosher...   
1  ['2 large egg whites', '1 pound new potatoes (...   
2  ['1 cup evaporated milk', '1 cup whole milk', ...   
3  ['1 (¾- to 1-pound) round Italian loaf, cut in...   
4  ['1 teaspoon dark brown sugar', '1 teaspoon ho...   
5  ['2 chamomile tea bags', '1½ oz. reposado tequ...   
6  ['3 oz. Grand Marnier', '1 oz. Amaro Averna', ...   
7  ['¼ cup granulated sugar', '¾ tsp. ground turm...   
8  ['¾ cup assorted dals (such as chana dal, moon...   
9  ['1 (14.5-ounce) can basic lentil soup, like A...   

                                        Instructions  \
0  Pat chicken dry with paper towels, season all ...   
1  Preheat oven to 400°F and line a rimmed baking...   
2  Place a rack in middle of oven; preheat to 400...   
3  Preheat oven to 350°F with rack in middle. Gen...   
4  Stir together brown sugar and hot water in a c...   
5  Place 2 chamomile tea bags in a heatsafe vesse...   
6  Add 3 oz. Grand Marnier, 1 oz. Amaro Averna, a...   
7  For the turmeric syrup, combine ½ cup hot wate...   
8  Combine dals, rice, and barley in a medium bow...   
9  Place an oven rack in the center of the oven, ...   

                                          Image_Name  \
0  miso-butter-roast-chicken-acorn-squash-panzanella   
1         crispy-salt-and-pepper-potatoes-dan-kluger   
2         thanksgiving-mac-and-cheese-erick-williams   
3          italian-sausage-and-bread-stuffing-240559   
4                 newtons-law-apple-bourbon-cocktail   
5               warm-comfort-tequila-chamomile-toddy   
6                    apples-and-oranges-spiked-cider   
7                  turmeric-hot-toddy-claire-sprouse   
8                            instant-pot-lamb-haleem   
9     spiced-lentil-and-caramelized-onion-baked-eggs   

                                 Cleaned_Ingredients  
0  ['1 (3½–4-lb.) whole chicken', '2¾ tsp. kosher...  
1  ['2 large egg whites', '1 pound new potatoes (...  
2  ['1 cup evaporated milk', '1 cup whole milk', ...  
3  ['1 (¾- to 1-pound) round Italian loaf, cut in...  
4  ['1 teaspoon dark brown sugar', '1 teaspoon ho...  
5  ['2 chamomile tea bags', '1½ oz. reposado tequ...  
6  ['3 oz. Grand Marnier', '1 oz. Amaro Averna', ...  
7  ['¼ cup granulated sugar', '¾ tsp. ground turm...  
8  ['¾ cup assorted dals (such as chana dal, moon...  
9  ['1 (14.5-ounce) can basic lentil soup, like A...

(13501, 5)


### Data Cleaning and Sanity check up

1. Check for duplicates and dropping them
2. checking for missing values, drop them or fill them
3. dropping ingredents column and renaming and rearrange the cleaned_ingredients to ingredients

- Check for duplicates

In [10]:
# checking shape of our data
df.shape

(13501, 5)

In [11]:
print(f'Number of duplicates: {df.duplicated().sum()}')

Number of duplicates: 0


In [12]:
#drop duplicated rows
df.drop_duplicates(inplace=True)
print(f'Number of duplicates after dropping: {df.duplicated().sum()}')

Number of duplicates after dropping: 0


- Check for missing values

In [13]:
df.isnull().sum().sort_values(ascending=False)

Instructions           8
Title                  5
Ingredients            0
Image_Name             0
Cleaned_Ingredients    0
dtype: int64

In [14]:
#drop rows with missing values
df.dropna(inplace=True)
print(f'Number of missing values after dropping: {df.isnull().sum().sum()}')

Number of missing values after dropping: 0


- There is the Ingredients and Cleaned Ingredients column, check if there is any difference between the two.

In [15]:
df['Ingredients'][5]

"['2 chamomile tea bags', '1½ oz. reposado tequila', '¾ oz. fresh lemon juice', '1 Tbsp. agave nectar']"

In [16]:
df['Cleaned_Ingredients'][5]

"['2 chamomile tea bags', '1½ oz. reposado tequila', '¾ oz. fresh lemon juice', '1 Tbsp. agave nectar']"

- There's no significant difference between Ingredients and cleaned Ingredients. Thus, we drop the Ingredients column and rename cleaned_ingredients  ingredients.

In [17]:
#move the cleaned_ingredients column to the second column
df = df[['Title', 'Cleaned_Ingredients', 'Ingredients', 'Instructions', 'Image_Name']]

#drop the ingredients column
df = df.drop(columns=['Ingredients','Image_Name'])


In [18]:
# rename the cleaned ingredients column
df = df.rename(columns={'Cleaned_Ingredients':'Ingredients'})
df.head()


Title  \
0  Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                    Crispy Salt and Pepper Potatoes   
2                        Thanksgiving Mac and Cheese   
3                 Italian Sausage and Bread Stuffing   
4                                       Newton's Law   

                                         Ingredients  \
0  ['1 (3½–4-lb.) whole chicken', '2¾ tsp. kosher...   
1  ['2 large egg whites', '1 pound new potatoes (...   
2  ['1 cup evaporated milk', '1 cup whole milk', ...   
3  ['1 (¾- to 1-pound) round Italian loaf, cut in...   
4  ['1 teaspoon dark brown sugar', '1 teaspoon ho...   

                                        Instructions  
0  Pat chicken dry with paper towels, season all ...  
1  Preheat oven to 400°F and line a rimmed baking...  
2  Place a rack in middle of oven; preheat to 400...  
3  Preheat oven to 350°F with rack in middle. Gen...  
4  Stir together brown sugar and hot water in a c...

In [19]:
df['Ingredients'][5]

"['2 chamomile tea bags', '1½ oz. reposado tequila', '¾ oz. fresh lemon juice', '1 Tbsp. agave nectar']"

### DataFrame two
The dataset is CSV file, it contains rows 142 and 16 columns respectively as follows:
1. Title: Represents the Title of the Food Dish.
2. Ingredients: Contains the ingredients as they were scraped from the website.
3. Instructions: Has the recipe instructions to be followed to recreate the dish.

In [20]:
#explore the recipeslmp.csv file
df2 = pd.read_csv("Data\RecipesImp.csv")
display(df2.head())

#display the shape
print(df2.shape)

title  index  page  \
0                           Kaimati(Fried Dumplings)  15003    24   
1                      Mahamri\r\n(Swahili Doughnut)  15004    26   
2  Enriched Mandazi \r\n(East African Doughnuts, ...  15124    28   
3  Basic Mandazi \r\n(East African Doughnuts, Basic)  15125    30   
4                  Meat Samosa\r\n(Sambusa ya Nyama)  15025    32   

                                               about  \
0  Kaimatis get their unique flavour from the sty...   
1  This is a typical traditional recipe among the...   
2  A popular snack among urban dwellers across th...   
3  You will find this recipe in any home across K...   
4  Nothing more delicious like the Kenyan meaty s...   

                                         ingridients  \
0   wheat flour, refined\r\nwater, vanilla essenc...   
1  wheat flour,\r\ncoconut milk\r\nwhite sugar\r\...   
2  self-raising wheat flour\r\neggs\r\nmargarine\...   
3  all-purpose wheat flour\r\nbaking powder\r\nsu...   
4  minced beef\r\ncoriander, fresh\r\nleek\r\ngar...   

                                         preparation  \
0   Put yeast in a small container.\r\n Add 50ml ...   
1    Break the coconut shell, drain the water and...   
2  ?  Put flour, salt, sugar and lemon rind into ...   
3  ?  Put the wheat flour into a bowl, add baking...   
4  ?  Put the meat in a pan over a fire. Stir con...   

                        nutrition per 100g of recipe  energy(kcal)  fat(g)  \
0  Energy 1,795 kJ/ 429 kcal | Fat 21.8 g | Carbo...         429.0    21.8   
1  Energy 1,728 kJ/ 413 kcal | Fat 22.1 g | Carbo...         413.0    22.1   
2  Energy 1,590 kJ/ 379 kcal | Fat 16.1 g | Carbo...         379.0    16.1   
3  Energy 1,430kJ/ 340 kcal | Fat 12.9 g | Carboh...         340.0    12.9   
4  Energy 1,854 kJ/ 443 kcal | Fat 22.2 g | Carbo...         443.0    22.2   

   carbohydrates(g)  proteins(g)  fibre(g) vitamin A(mcg)  iron(mg)  zinc(mg)  \
0              52.8          4.6       1.6             30       2.1      0.45   
1              46.6          6.0       2.1             41       2.8      0.56   
2              49.9          7.6       2.2             90       3.3      0.66   
3              48.7          6.4       2.1             48       3.5      0.52   
4              40.5         18.8       3.1             66      11.5      2.99   

   F_factor_est  
0           0.4  
1           0.4  
2           0.4  
3           0.4  
4           0.4

(142, 16)


- Since we want only a few columns to recommend the possible recipes, we need to drop some columns.

In [21]:
df2.columns


Index(['title', 'index', 'page', 'about', 'ingridients', 'preparation',
       'nutrition per 100g of recipe', 'energy(kcal)', 'fat(g)',
       'carbohydrates(g)', 'proteins(g)', 'fibre(g)', 'vitamin A(mcg)',
       'iron(mg)', 'zinc(mg)', 'F_factor_est'],
      dtype='object')

### Data Cleaning and Sanity check up

1. Check for duplicates and dropping them
2. checking for missing values, drop them or fill them
3. dropping irrelevant columns

In [22]:
columns_to_keep = ['title','ingridients','preparation']

df2 = df2[columns_to_keep]
df2.head()

title  \
0                           Kaimati(Fried Dumplings)   
1                      Mahamri\r\n(Swahili Doughnut)   
2  Enriched Mandazi \r\n(East African Doughnuts, ...   
3  Basic Mandazi \r\n(East African Doughnuts, Basic)   
4                  Meat Samosa\r\n(Sambusa ya Nyama)   

                                         ingridients  \
0   wheat flour, refined\r\nwater, vanilla essenc...   
1  wheat flour,\r\ncoconut milk\r\nwhite sugar\r\...   
2  self-raising wheat flour\r\neggs\r\nmargarine\...   
3  all-purpose wheat flour\r\nbaking powder\r\nsu...   
4  minced beef\r\ncoriander, fresh\r\nleek\r\ngar...   

                                         preparation  
0   Put yeast in a small container.\r\n Add 50ml ...  
1    Break the coconut shell, drain the water and...  
2  ?  Put flour, salt, sugar and lemon rind into ...  
3  ?  Put the wheat flour into a bowl, add baking...  
4  ?  Put the meat in a pan over a fire. Stir con...

In [23]:
#clean the column names
#change the ingridient column name to ingredients
df2.rename(columns={'ingridients':'ingredients','preparation':'instructions'}, inplace=True)

#capitalize the column names
df2.columns = df2.columns.str.capitalize()

# Function to process the Ingredients column
def process_ingredients(ingredients):
    # Remove unwanted characters and split by commas
    return [ingredient.strip() for ingredient in ingredients.replace('[','').replace(']','').replace("'", "").replace('\n', ',').split(',')]

# Apply the function to each DataFrame
df['Ingredients'] = df['Ingredients'].apply(lambda x: [ingredient.strip() for ingredient in x.split(',')])
df2['Ingredients'] = df2['Ingredients'].apply(process_ingredients)




In [24]:
# Check the processed DataFrames
df[['Title', 'Ingredients']]


Title  \
0      Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                        Crispy Salt and Pepper Potatoes   
2                            Thanksgiving Mac and Cheese   
3                     Italian Sausage and Bread Stuffing   
4                                           Newton's Law   
...                                                  ...   
13496                               Brownie Pudding Cake   
13497  Israeli Couscous with Roasted Butternut Squash...   
13498  Rice with Soy-Glazed Bonito Flakes and Sesame ...   
13499                                        Spanakopita   
13500  Mexican Poblano, Spinach, and Black Bean "Lasa...   

                                             Ingredients  
0      [['1 (3½–4-lb.) whole chicken', '2¾ tsp. koshe...  
1      [['2 large egg whites', '1 pound new potatoes ...  
2      [['1 cup evaporated milk', '1 cup whole milk',...  
3      [['1 (¾- to 1-pound) round Italian loaf, cut i...  
4      [['1 teaspoon dark brown sugar', '1 teaspoon h...  
...                                                  ...  
13496  [['1 cup all-purpose flour', '2/3 cup unsweete...  
13497  [['1 preserved lemon', '1 1/2 pound butternut ...  
13498  [['Leftover katsuo bushi (dried bonito flakes)...  
13499  [['1 stick (1/2 cup) plus 1 tablespoon unsalte...  
13500  [['12 medium to large fresh poblano chiles (2 ...  

[13493 rows x 2 columns]

In [25]:
df2[['Title', 'Ingredients']]

Title  \
0                             Kaimati(Fried Dumplings)   
1                        Mahamri\r\n(Swahili Doughnut)   
2    Enriched Mandazi \r\n(East African Doughnuts, ...   
3    Basic Mandazi \r\n(East African Doughnuts, Basic)   
4                    Meat Samosa\r\n(Sambusa ya Nyama)   
..                                                 ...   
137                   Bhature\r\n (Fried Indian Bread)   
138  Vimumunya vya \r\nSukari\r\n (Sweetened Pumpki...   
139                         Siro\r\n (Semolina & Nuts)   
140                         Chaas\r\n(Diluted Yoghurt)   
141                                    Groundnut Sauce   

                                           Ingredients  
0    [wheat flour, refined, water, vanilla essence,...  
1    [wheat flour, , coconut milk, white sugar, dry...  
2    [self-raising wheat flour, eggs, margarine, Ri...  
3    [all-purpose wheat flour, baking powder, sugar...  
4    [minced beef, coriander, fresh, leek, garlic, ...  
..                                                 ...  
137  [wheat flour, salt, sugar, ghee, cooking oil, ...  
138    [pumpkin, cardamon, sugar, coconut milk, water]  
139  [semolina flour, cow ghee, cow milk, sugar, pi...  
140                   [natural yoghurt, water, salt, ]  
141                [groundnut, salt, sour milk, water]  

[142 rows x 2 columns]

- Chek for missing values

In [26]:
print(f'number of missing values: {df2.isnull().sum().sum()}')

number of missing values: 0


### Merging our datasets
- Since both dataframes have no missing values and duplicates we can merge them now. The new dataset is knowns as Combined_df

In [27]:
#merge the two dataframes
combined_df = pd.concat([df,df2])
#check the shapes of the three dfs
print(f'Dataframe 1 has a shape of: {df.shape}')
print(f'Dataframe 2 has a shape of: {df2.shape}')
print(f'Combined dataframe has a shape of: {combined_df.shape}')

#reset the index
combined_df = combined_df.reset_index(drop=True)



Dataframe 1 has a shape of: (13493, 3)
Dataframe 2 has a shape of: (142, 3)
Combined dataframe has a shape of: (13635, 3)


### Clean the Combined DataFrame

1. Remove newlines and extra spaces
2. Convert string ingredients to lists
3. Clean the Instructions Column

In [28]:
combined_df.head()

Title  \
0  Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                    Crispy Salt and Pepper Potatoes   
2                        Thanksgiving Mac and Cheese   
3                 Italian Sausage and Bread Stuffing   
4                                       Newton's Law   

                                         Ingredients  \
0  [['1 (3½–4-lb.) whole chicken', '2¾ tsp. koshe...   
1  [['2 large egg whites', '1 pound new potatoes ...   
2  [['1 cup evaporated milk', '1 cup whole milk',...   
3  [['1 (¾- to 1-pound) round Italian loaf, cut i...   
4  [['1 teaspoon dark brown sugar', '1 teaspoon h...   

                                        Instructions  
0  Pat chicken dry with paper towels, season all ...  
1  Preheat oven to 400°F and line a rimmed baking...  
2  Place a rack in middle of oven; preheat to 400...  
3  Preheat oven to 350°F with rack in middle. Gen...  
4  Stir together brown sugar and hot water in a c...

In [29]:
combined_df[-10:]

Title  \
13625           Vinolo\r\n(Banana and Maize Flour Ugali)   
13626                      Finger Millet \r\nFlour Ugali   
13627                                      White Chapati   
13628                                      Brown Chapati   
13629                          Roti \r\n(Indian Chapati)   
13630                   Bhature\r\n (Fried Indian Bread)   
13631  Vimumunya vya \r\nSukari\r\n (Sweetened Pumpki...   
13632                         Siro\r\n (Semolina & Nuts)   
13633                         Chaas\r\n(Diluted Yoghurt)   
13634                                    Groundnut Sauce   

                                             Ingredients  \
13625                 [banana green, maize flour, water]   
13626                             [finger millet, water]   
13627     [wheat flour, water, sugar, salt, cooking oil]   
13628   [wheat flour, water, sugar, , salt, cooking oil]   
13629  [wheat flour, salt, water, cooking oil, cow ghee]   
13630  [wheat flour, salt, sugar, ghee, cooking oil, ...   
13631    [pumpkin, cardamon, sugar, coconut milk, water]   
13632  [semolina flour, cow ghee, cow milk, sugar, pi...   
13633                   [natural yoghurt, water, salt, ]   
13634                [groundnut, salt, sour milk, water]   

                                            Instructions  
13625  Preparation 5 minutes | Cooking 40 minutes | \...  
13626  Preparation time 5 minutes | Cooking time 15 m...  
13627  Preparation 30 minutes | Cooking 30 minutes | ...  
13628  Preparation 30 minutes | Cooking 30 minutes | ...  
13629  Preparation 3 hours | Cooking 21 minutes | Ser...  
13630  Preparation 1 hour 15 minutes | Cooking 30 min...  
13631  Preparation 5 minutes | Cooking 45 minutes | \...  
13632  Preparation 15 minutes | Cooking 30 minutes | ...  
13633  Preparation 5 minutes | Serves 2\r\n?Add natur...  
13634  Preparation 5 minutes | Cooking 1 hour 40 minu...

In [30]:
#check for missing values
print(combined_df.isnull().sum().sort_values(ascending=False))


Title           0
Ingredients     0
Instructions    0
dtype: int64


- Clean the columns of the combined_df

In [31]:
#write a function to clean the columns
"""
This function should:
1. Clean the Title Column: Remove newlines and extra spaces
2. Clean the Ingredients Column: Convert string ingredients to lists: If they are not already in a list format, convert them and clean any extra spaces.
                                Ensure there are no empty strings or duplicates within each list of ingredients.
3. Clean the Instructions Column: Similar to the Title, ensure that the instructions are clean and properly formatted.
"""

def clean_combined_df(df):
    #title column
    # Remove newlines and extra spaces
    df['Title'] = df['Title'].str.replace('\n', '').str.strip()
    #Ingredients column
    # Remove empty ingredients from each list of ingredients
    df['Ingredients'] = df['Ingredients'].apply(lambda x: [ingredient.strip() for ingredient in x if ingredient.strip() != ''])

    # If there are any duplicate ingredients in each list, remove them
    df['Ingredients'] = df['Ingredients'].apply(lambda x: list(set(x)))

    # Clean the Instructions column
    df['Instructions'] = df['Instructions'].str.replace('\n', '') \
                                       .str.replace('?', '.') \
                                       .str.replace('|', ',') \
                                       .str.replace('\r', '') \
                                       .str.strip()
    #remove   special characters like \x02 from words like  gradu\x02ally
    df['Instructions'] = df['Instructions'].apply(lambda x: re.sub(r'[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\xff]', '', x))#remove special characters
    #if a sentence has more than one fullstop, replace it with one
    df['Instructions'] = df['Instructions'].apply(lambda x: re.sub(r'\.{2,}', '.', x))
    return df

# Use the function on your DataFrame
combined_cleaned = clean_combined_df(combined_df)
combined_cleaned

Title  \
0      Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                        Crispy Salt and Pepper Potatoes   
2                            Thanksgiving Mac and Cheese   
3                     Italian Sausage and Bread Stuffing   
4                                           Newton's Law   
...                                                  ...   
13630                     Bhature\r (Fried Indian Bread)   
13631  Vimumunya vya \rSukari\r (Sweetened Pumpkin & ...   
13632                           Siro\r (Semolina & Nuts)   
13633                           Chaas\r(Diluted Yoghurt)   
13634                                    Groundnut Sauce   

                                             Ingredients  \
0      ['2 small acorn squash (about 3 lb. total)', m...   
1      ['1 pound new potatoes (about 1 inch in diamet...   
2      ['1 cup whole milk', '1 tsp. onion powder', pl...   
3      ['1 stick unsalted butter, cut into pieces', '...   
4      ['2 teaspoons apple butter (storebought or hom...   
...                                                  ...   
13630  [fenugreek leaves, natural yoghurt, sugar, wat...   
13631    [coconut milk, water, sugar, pumpkin, cardamon]   
13632  [cow ghee, semolina flour, cow milk, sugar, al...   
13633                     [water, salt, natural yoghurt]   
13634                [water, sour milk, salt, groundnut]   

                                            Instructions  
0      Pat chicken dry with paper towels, season all ...  
1      Preheat oven to 400F and line a rimmed baking ...  
2      Place a rack in middle of oven; preheat to 400...  
3      Preheat oven to 350F with rack in middle. Gene...  
4      Stir together brown sugar and hot water in a c...  
...                                                  ...  
13630  Preparation 1 hour 15 minutes , Cooking 30 min...  
13631  Preparation 5 minutes , Cooking 45 minutes , S...  
13632  Preparation 15 minutes , Cooking 30 minutes , ...  
13633  Preparation 5 minutes , Serves 2.Add natural y...  
13634  Preparation 5 minutes , Cooking 1 hour 40 minu...  

[13635 rows x 3 columns]

In [32]:
combined_cleaned['Instructions'][13634]

'Preparation 5 minutes , Cooking 1 hour 40 minutes , Serves 4.Place a saucepan over fire and let it preheat.Add the groundnuts, salt and 1/2 a cup of water oras desired.Cook until the water evaporates as you stir gradually. When ready, the nuts produce a pop sound.Once they pop, turn down the heat and continue stirring until the groundnuts are dry (about 13minutes).Remove from heat and allow it to cool down.Using a blender, blend the nuts into a paste. Apestle and mortar can be used in the absence of ablender.Put the groundnut paste into a bowl, add sour milkand stir into thick paste. Water or fresh milk can beused in place of the sour milk.Once ready, put another pan on the heat, add thepeanut paste and stir.Stir until it is smooth but not too thick.Serve hot with green leafy vegetables of yourchoice, fish, sweet potatoes, green bananas, ugali,etc'

## Modelling

### Function to get Recipes based on Ingredients
1. Applying tokenization in our ingredient column
2. Defining Function for Ingredients and Title, that is the system is set to work and recommend based on the input, the user can input either title or the ingredients to get the instruction and title or ingredients which will depend with his input
3. Feature Engineering
4. Intiating TF-IDF Vectorizer to  create vectors based on both the recipe title and ingredients.

- Let's see if we can tokenize our Ingredients and see if the system recommends recipes based on the ingredients

In [33]:

def clean_ingredients(ingredients_list):
    # Remove any extra single quotes and fix formatting for each ingredient
    cleaned_list = [re.sub(r"['\"]", "", ingredient) for ingredient in ingredients_list]  # Remove quotes
    cleaned_list = [re.sub(r'\s+', ' ', ingredient) for ingredient in cleaned_list]  # Normalize spaces
    return cleaned_list

# Apply the cleaning function
combined_df['Ingredients'] = combined_df['Ingredients'].apply(clean_ingredients)


# Function to tokenize and normalize ingredients
def tokenize_and_normalize(ingredients_list):
    tokens = []
    for ingredient in ingredients_list:
        # Split ingredient string by commas and strip whitespace
        split_ingredients = [i.strip().lower() for i in ingredient.split(',')]
        
        # Further clean each token: remove unwanted characters
        split_ingredients = [re.sub(r'[^\w\s]', '', i) for i in split_ingredients]  # Remove punctuation
        split_ingredients = [re.sub(r'\s+', ' ', i) for i in split_ingredients]  # Normalize whitespace
        
        # Extend the tokens list with cleaned ingredients
        tokens.extend(split_ingredients)
    
    return tokens


# Apply the function to the Ingredients column
combined_df['Ingredients'] = combined_df['Ingredients'].apply(tokenize_and_normalize)


In [34]:
def recommend_recipes(input_ingredients, combined_df):
    # Normalize user input by stripping whitespace and converting to lower case
    input_ingredients = [ingredient.strip().lower() for ingredient in input_ingredients.split(',')]
    
    # Find matching recipes
    matched_recipes = combined_df[combined_df['Ingredients'].apply(lambda x: any(ingredient in x for ingredient in input_ingredients))]
    
    # Check if any recipes were found
    if matched_recipes.empty:
        return "No recipe found. Try again."
    
    return matched_recipes[['Title','Ingredients', 'Instructions']]


- When the user inputs an ingredient:

In [35]:
# Example user input ingredients
user_input = "thyme"

# Get recommendations
recommended_recipes = recommend_recipes(user_input, combined_df)

# Display recommendations
recommended_recipes


Title  \
273               Braised Celery With Lentils and Garlic   
1479                                   Kale or Chard Pie   
1586                                    Make-Ahead Gravy   
1897      3-Ingredient Garlic-Herb Grilled Chicken Wings   
2766                         Herby Ricotta-Poblano Tacos   
5510                                        Grilled Fish   
5988                                            Gazpacho   
6253                    Duck Breast & Zucchini Tournedos   
6851                                 Chicken & Rice Soup   
7203        Grilled Chicken with Almond and Garlic Sauce   
7808   Rib-Eye, T-Bone, and Strip Steaks Cut Over One...   
8057                                            Gazpacho   
8451                    Heirloom Tomato Herb Pasta Salad   
11968  Sliced Filet Mignon with Fava Beans, Radishes,...   
12036                                Fresh Herb Spaetzle   

                                             Ingredients  \
273    [a small handful hardy herb sprigs such as ros...   
1479   [thyme, andor chives, 6 eggs, 2 tablespoons bu...   
1586   [1 large onion, 1 celery stalk, thyme, 14 cup ...   
1897   [thyme, andor oregano, plus more, 14 cup chopp...   
2766   [big, 4 teaspoons chopped fresh herbs such as ...   
5510   [about 3 cloves, 12 teaspoon ground cumin, or ...   
5988   [1 large cucumber, juice of 12 a lemon, 1 red ...   
6253   [rosemary, 3 young zucchini, 1 34 oz dried cep...   
6851   [thyme, skinless, celery 1 stalk, diced, or pa...   
7203   [12 cup ground almonds, such as parsley, 3 tab...   
7808   [canola oil or vegetable oil, 4 tbsp unsalted ...   
8057   [12 cup thinly sliced chives or scallion green...   
8451   [13 cup crème fraîche, 2 pounds mixed heirloom...   
11968  [thyme, 10 medium radishes, 1 tablespoon butte...   
12036  [thyme, and chives, rosemary, divided, 4 table...   

                                            Instructions  
273    Heat  cup oil in a large high-sided skillet ov...  
1479   Heat the oven to 375F. Melt the butter in a la...  
1586   Heat oil in a large saucepan over medium-high....  
1897   Pat wings very dry with paper towels. Whisk ga...  
2766   On an ungreased griddle or small, heavy skille...  
5510   Combine all the ingredients except fish and le...  
5988   1. In a bowl, reserve 2 tablespoons each of th...  
6253   For the mushroom powder, process the dried cep...  
6851   Combine the broth, rice, and 2 cups (16 fl oz/...  
7203   1. For the sauce, soak the beans in a generous...  
7808   Generously season the steaks with salt and pep...  
8057   1. Reserve 2 tablespoons each of the tomatoes,...  
8451   Whisk together oil, crme frache, vinegar, suga...  
11968  Whisk vinegar and mustard in small bowl. Gradu...  
12036  Blend flour, salt, pepper, and nutmeg in large...

- The results are titles, ingredients and instructions all with the name thyme in it.

- When the user enters a title:

In [36]:
# Example user input ingredients
user_input = "Gazpacho"

# Get recommendations
recommended_recipes = recommend_recipes(user_input, combined_df)

# Display recommendations
recommended_recipes


'No recipe found. Try again.'

In [37]:
#check if there is a recipe with the title 'Gazpacho'
combined_df[combined_df['Title'].str.contains('Gazpacho', case=False)]

Title  \
133                               Speedy Summer Gazpacho   
1771                              Sippin’ Green Gazpacho   
3894                                  Gochujang Gazpacho   
4089                                        Wee Gazpacho   
4476                 Watermelon Gazpacho with Feta Crema   
5297                               Creamy Green Gazpacho   
5988                                            Gazpacho   
6163                     Mixed Berry Gazpacho with Basil   
6260                  Stone Fruit Gazpacho with Scallops   
6564         Fiery Grilled Shrimp with Honeydew Gazpacho   
6765                                            Gazpacho   
7585   Grilled Heirloom Tomato and Mozzarella Sandwic...   
8057                                            Gazpacho   
8230                                Green Gazpacho Shots   
8449                       Tomato and Tomatillo Gazpacho   
8509          Seared Mahi-Mahi with Green Gazpacho Sauce   
8754                            Tomato Cucumber Gazpacho   
8856             Cucumber Gazpacho with Shrimp and Melon   
9405                                  Farmstand Gazpacho   
9415                                      Gazpacho Salsa   
9805                           Tomatillo-Pepita Gazpacho   
10722                Grilled Tomato-Bell Pepper Gazpacho   
10765       Ziti with Grilled-Gazpacho Sauce and Sausage   
11254                              Gazpacho Bloody Marys   
11674                              Celery Apple Gazpacho   
11699                 Grilled Gazpacho Salad with Shrimp   
12499                                    Speedy Gazpacho   

                                             Ingredients  \
133    [1 celery stalk, grated, ½ oz basil leaves, 2 ...   
1771   [2 cups coarsely chopped arugula, 2 pounds eng...   
3894   [2 tablespoons gochujang korean hot pepper pas...   
4089   [1 teaspoon lemon juice, yellow or orange, 1 t...   
4476   [cut into 12 pieces about 2 cups, seeds remove...   
5297   [peeled and cut into large chunks, or more to ...   
5988   [1 large cucumber, juice of 12 a lemon, 1 red ...   
6163   [1 tablespoon fresh orange juice, 1 teaspoon f...   
6260   [1 pound white peaches, pitted, espelette pepp...   
6564   [1 small fennel bulb, 12 english cucumber, see...   
6765   [2 14 lb ripe tomatoes, 2 onions, 4 oz crouton...   
7585   [cut into 14inch slices, 1 pound ripe green he...   
8057   [12 cup thinly sliced chives or scallion green...   
8230   [trimmed and coarsely chopped, husks removed, ...   
8449   [husked, divided, 1 cup water, 12 pound fresh ...   
8509   [5 12 tablespoons olive oil, divided, 1 12 tab...   
8754   [divided, 3 small garlic cloves, 5 cups coarse...   
8856   [12 cup loosely packed assorted fresh herbs, p...   
9405   [2 dashes tabasco sauce, 2 cups peeled and dic...   
9415   [12 cup finely chopped red onion, cut into 14i...   
9805   [husked, 1 12ounce container cherry tomatoes, ...   
10722  [12 teaspoon ground cumin, and from tiendacom,...   
10765  [divided, 3 bell peppers any color, 12 cup cho...   
11254  [4 teaspoons worcestershire sauce, chopped abo...   
11674  [8 to 9 celery ribs, 14 cup blanched almonds, ...   
11699  [1 red bell pepper, halved lengthwise, 2 6x3x1...   
12499  [cut into large chunks, cut into chunks, halve...   

                                            Instructions  
133    Put all the ingredients in a food processor an...  
1771   Pure cucumbers, garlic, and 1/2 cup water in a...  
3894   Pulse 1 1/2 cups tomato, 1/4 cup cucumber, ora...  
4089   Place all the ingredients in a food processor ...  
4476   Pure watermelon, tomato, cucumber, jalapeo, oi...  
5297   Reserve one-quarter of the tomato, two cucumbe...  
5988   1. In a bowl, reserve 2 tablespoons each of th...  
6163   Combine first 6 ingredients and basil sprig in...  
6260   Combine the peaches, plums, watermelon, garlic...  
6564   Cut away the honeydew rind and the dark green ...  
6765   Peel and cut each garlic clove in

- They get the response "No recipe found.Try again" but if you check among the foods, there is a recipe for Gazpacho.

- In the above function we see that when a user enters an ingredient it results in sample recipes the user can cook. However when the user tries to enter a title of a food they want, it returns the response "No recipe found. Try Again."
- If you search the title of the food in the list of food, it's there. So the function isn't really working accordingly.
- Thus we can try Feature Engineering. Create a new column with combined title and ingredients, Vectorize, create cosine similarities and see how well it performs compared to our function above.

### Feature Engineering

- Create the column containing combined title and ingredient.

In [38]:
# Combining title and ingredients
"""
This code is meant to combine both the ingredients and the title into a 
single string for each recipe this will allow to system to search both 
the ingredients and the title since ther user can input any of them.
"""
combined_df['Title_Ingredients'] = combined_df['Title'] + ' ' + combined_df['Ingredients'].apply(lambda x: ' '.join(x))
combined_df.head()

Title  \
0  Miso-Butter Roast Chicken With Acorn Squash Pa...   
1                    Crispy Salt and Pepper Potatoes   
2                        Thanksgiving Mac and Cheese   
3                 Italian Sausage and Bread Stuffing   
4                                       Newton's Law   

                                         Ingredients  \
0  [2 small acorn squash about 3 lb total, melted...   
1  [1 pound new potatoes about 1 inch in diameter...   
2  [1 cup whole milk, 1 tsp onion powder, plus mo...   
3  [1 stick unsalted butter, cut into pieces, ¾ c...   
4  [2 teaspoons apple butter storebought or homem...   

                                        Instructions  \
0  Pat chicken dry with paper towels, season all ...   
1  Preheat oven to 400F and line a rimmed baking ...   
2  Place a rack in middle of oven; preheat to 400...   
3  Preheat oven to 350F with rack in middle. Gene...   
4  Stir together brown sugar and hot water in a c...   

                                   Title_Ingredients  
0  Miso-Butter Roast Chicken With Acorn Squash Pa...  
1  Crispy Salt and Pepper Potatoes 1 pound new po...  
2  Thanksgiving Mac and Cheese 1 cup whole milk 1...  
3  Italian Sausage and Bread Stuffing 1 stick uns...  
4  Newton's Law 2 teaspoons apple butter storebou...

In [39]:
# preview combined column of title together with ingredients
combined_df['Title_Ingredients'][0]

'Miso-Butter Roast Chicken With Acorn Squash Panzanella 2 small acorn squash about 3 lb total melted ¼ cup allpurpose flour cut into 1 pieces kosher salt 2 cups unsalted chicken broth divided freshly ground black pepper 3 tbsp apple cider vinegar room temperature 2 medium apples such as gala or pink lady about 14 oz total plus 3 tbsp room temperature 2 tsp white miso 2 tbsp extravirgin olive oil pinch of crushed red pepper flakes freshly ground pepper ⅓ loaf goodquality sturdy white bread 2 tbsp unsalted butter 6 tbsp unsalted butter ½ small red onion ¼ tsp ground allspice cored ¼ cup dry white wine 2 tbsp finely chopped sage thinly sliced plus more 1 3½4lb whole chicken 1 tbsp finely chopped rosemary 2¾ tsp kosher salt torn into 1 pieces about 2½ cups 1 tbsp white miso'

### TF-IDF Vectorizer
-The TF-IDF vectorizer will then create vectors based on both the recipe title and ingredients. This will give you a more comprehensive representation of the recipe.

In [40]:
# intiliazing TFIDF Vectorizer
tfidf_vectorizer = TfidfVectorizer()
# fit the vectorizer
tfidf_matrix = tfidf_vectorizer.fit_transform(combined_df['Title_Ingredients'])


- Check the shape of the matrix and see the features to know how the vectorization has taken place.

In [41]:
# Check the shape of the TF-IDF matrix (rows x features)
print(f'Shape of matrix:{tfidf_matrix.shape}')

# Get the feature names (terms)
tfidf_feature_names = tfidf_vectorizer.get_feature_names_out()
print(f'Feature Names:{tfidf_feature_names[:10]}')  # Show first 10 features (terms)

Shape of matrix:(13635, 11974)
Feature Names:['00' '000mg' '011' '018' '018oz' '02' '025' '028' '035' '037']


- In the above we can see how the shape of the matrix is 13,635 by 13,690 after vectorization and sample feature names.

- Let's get the cosine Similarities

In [42]:
#cosine similarity
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

#print the cosine similarity matrix shape
print(f'Cosine Similarity Matrix Shape: {cosine_sim.shape}')


Cosine Similarity Matrix Shape: (13635, 13635)


- Let's see how well our system recommends recipes based on the cosine similarities.
- Let's take a look at the first recipe and the top 5 recipes with highest cosine similarities with it.

In [43]:
#example getting top 5 most similar recipes with the first recipe
similar_recipes = cosine_sim[0].argsort()[-6:-1] #


In [44]:
#display the first recipe
combined_df.iloc[0][['Title', 'Ingredients', 'Instructions']]

Title           Miso-Butter Roast Chicken With Acorn Squash Pa...
Ingredients     [2 small acorn squash about 3 lb total, melted...
Instructions    Pat chicken dry with paper towels, season all ...
Name: 0, dtype: object

In [45]:
# Display the similar recipes
combined_df.iloc[similar_recipes][['Title', 'Ingredients', 'Instructions']][::-1]

Title  \
590   Grilled Whole Cauliflower with Miso Mayo   
129      Grilled Potatoes With Red Miso Butter   
331           Risotto With Mushrooms and Thyme   
4237                     Bacon-Cheddar Muffins   
26                           Miso-Squash Ramen   

                                            Ingredients  \
590   [12 tsp or more kosher salt, 1 tbsp ketchup, 1...   
129   [2 tbsp chopped parsley, freshly ground black ...   
331   [cut into pieces, kosher salt, ½ large white o...   
4237  [1 12 tbsp bacon fat, 6 tbsp grated parmesan, ...   
26    [4 cups lowsodium vegetable broth, kosher salt...   

                                           Instructions  
590   Prepare a grill for medium-high heat. Sprinkle...  
129   Prepare a grill for medium-high heat. Bring po...  
331   Combine 1 Tbsp. salt and 10 cups water in a me...  
4237  Preheat oven to 400 with rack near top. Line t...  
26    Preheat oven to 400F. Whisk oil, miso, maple s...

In [46]:
#display the similarities of the first recipe with the top 5 similar recipes
cosine_sim[0][similar_recipes][::-1].round(4)

array([0.4927, 0.4799, 0.4646, 0.4638, 0.4568])

- In the above, there have been 5 top recipe recommendations  based on their cosine similarities with the first recipe. The first recipe having the highest similarity of 0.4927

- Clean the title.

In [47]:
# clean title column
def clean_input(title):
    # Normalize titles by removing punctuation, converting to lowercase, and stripping extra spaces
    title = re.sub(r'[^\w\s]', '', title)
    title = re.sub(r'\s+', ' ', title).strip().lower()
    return title
combined_df['Title'] = combined_df['Title'].apply(clean_input)

- Create a function to recommend the recipes.

In [48]:
#function to recommend recipes
def recommend_recipe(user_input, combined_df, tfidf_vectorizer, tfidf_matrix, n=10):
    # Clean the user input
    user_input_cleaned = clean_input(user_input)
    
    # Check if the user input matches a specific title
    matching_title = combined_df[combined_df['Title'].str.lower().str.contains(user_input_cleaned)]
    
    #check if there is a match by title
    if not matching_title.empty:
        return matching_title[['Title', 'Ingredients', 'Instructions']]
    
    # If no exact title match is found, proceed with TF-IDF based recommendation
    # Transform the user input
    user_input_transformed = tfidf_vectorizer.transform([user_input_cleaned])
    
    # Compute the cosine similarity between the user input and the recipes
    cosine_sim = cosine_similarity(user_input_transformed, tfidf_matrix)
    
    # Get the index of the top n most similar recipes
    similar_recipes = cosine_sim[0].argsort()[-n:][::-1]
    
    # Return the top n most similar recipes
    return combined_df.iloc[similar_recipes][['Title', 'Ingredients', 'Instructions']]
    

In [49]:
#Example usage
user_input = "eggs,tomatoes,onions"
recommendations = recommend_recipe(user_input, combined_df, tfidf_vectorizer, tfidf_matrix)

# View recommendations
recommendations

Title  \
13634                                    groundnut sauce   
4552          charred and raw corn with chile and cheese   
4550            grilled rib eye with paprika vinaigrette   
4549   grilled zucchini and leeks with walnuts and herbs   
4548            asparagus with bacon and hardboiled eggs   
4547                     mint and cuminspiced lamb chops   
4546                         moroccan chicken brochettes   
4545                                grilled sesame squid   
4544              plum tarts with honey and black pepper   
4543                       pickled carrots with tarragon   

                                             Ingredients  \
13634                [water, sour milk, salt, groundnut]   
4552   [with seeds, husked, freshly ground black pepp...   
4550   [freshly ground pepper, 2 tablespoons light br...   
4549   [13 cup walnuts, divided, 2 large zucchini abo...   
4548   [12 small shallot, 1 tablespoon pure maple syr...   
4547   [1 tablespoon ground cumin, 1 cup fresh flatle...   
4546   [chopped tomatoes, 2 pounds skinless, kosher s...   
4545   [vegetable oil for grilling, 1 12 pounds clean...   
4544   [cut into 12 wedges, 1 tablespoon honey, fresh...   
4543   [1 tablespoon kosher salt, 1 pound young carro...   

                                            Instructions  
13634  Preparation 5 minutes , Cooking 1 hour 40 minu...  
4552   Prepare grill for medium heat. Cut kernels fro...  
4550   Heat oil and paprika in a small saucepan over ...  
4549   Prepare grill for medium-high heat. Toast waln...  
4548   Cook bacon in a large skillet over medium heat...  
4547   Pulse onions, cilantro, parsley, mint, cumin, ...  
4546   Place garlic in a mortar; season with salt and...  
4545   Prepare grill for high heat; oil grate with ve...  
4544   Preheat oven to 425. Cut pastry into six 4" sq...  
4543   Pack carrots, shallot, garlic, chiles, tarrago...

In [52]:
#Example usage
user_input = "githeri"
recommendations = recommend_recipe(user_input, combined_df, tfidf_vectorizer, tfidf_matrix)

# View recommendations
recommendations

Title  \
13545     githeri fresh beans and maize   
13546        githeri stewed maize beans   
13547  githeri sauted fresh maize beans   

                                             Ingredients  \
13545  [water, bean pods, spring onion, green maize c...   
13546  [maize and beans, water, salt, cooking oil, on...   
13547  [kidney beans, water, green maize cobs with ke...   

                                            Instructions  
13545  Preparation 2 hours , Cooking 20 minutes , Ser...  
13546  Preparation 10 minutes , Cooking 15 minutes , ...  
13547  Preparation 5 hours 30 minutes , Cooking 10 mi...

- Based on the above, the system works in a way that when a user enters a specific title of a food like "mandazi" it returns all instances where the food title is seen otherwise, if the user enters ingredients, it converts the user input to a vector then performs the cosine similarity. Afterwards, it returns recipes of the top foods with highest cosine similarities.

- This system works accordingly in recommending the recipes to cook.


## Deployment

### Saving Code
- In order for a user to get recommended recipes, we need to save some code such as the recommend recipe function, combined_df, the tf_idf vectorizer as
.py and pickle files in order for a user to get feedback whenever they input data.
- So the TF-IDF Vectorizer and the combined dataframe will be saved as pickle files while the combined_df saved as .py file to make it easy to adjust here and there.

In [50]:
#saving the model
import joblib

#save the tfidf vectorizer
joblib.dump(tfidf_vectorizer, 'tfidf_vectorizer.pkl')

#save the tfidf matrix
joblib.dump(tfidf_matrix, 'tfidf_matrix.pkl')

#save the combined dataframe
combined_df.to_csv('combined_df.csv', index=False)
